rio_tiler/mosaic/methods/defaults.py

```diff
@@ -114,17 +114,86 @@ class StdevMethod(MosaicMethodBase):
     def __init__(self, enforce_data_type=True):
         """Overwrite base and init Stdev method."""
         super(StdevMethod, self).__init__()
+        self.enforce_data_type = enforce_data_type
         self.tile = []
 
     @property
     def data(self):
         """Return data and mask."""
         if self.tile:
             tile = numpy.ma.std(numpy.ma.stack(self.tile, axis=0), axis=0)
+            tile = tile.astype(self.tile[0].dtype) if self.enforce_data_type else tile
             return tile.data, ~tile.mask[0] * 255
         else:
             return None, None
 
     def feed(self, tile):
         """Add data to tile."""
         self.tile.append(tile)
+
+class GeoMedianMethod(MosaicMethodBase):
+
+    def __init__(self, enforce_data_type=True):
+        super(GeoMedianMethod, self).__init__()
+        self.enforce_data_type = enforce_data_type
+        self.tile = []
+
+    @property
+    def data(self):
+        """Return data and mask."""
+        if self.tile:
+            # Convert with rescaling factor to float32 for input to Dale's hdmedians
+            stacked_tile = (numpy.ma.stack(self.tile, axis=1) / 10000).astype(numpy.float32)
+            b, t, y, x = stacked_tile.shape
+            tile = numpy.ma.zeros((b, y, x)).astype(numpy.float32)
+
+            # TODO: How to implement geomedian without introducing too many new dependencies?
+            for xid in range(x):     # for each x-axis
+                for yid in range(y): # for each y-axis
+                    tile[:, yid, xid] = hd.geomedian(stacked_tile[:, :, yid, xid])
+
+            # Convert back to uint16
+            tile = (tile * 10000).astype(self.tile[0].dtype) if self.enforce_data_type else (tile * 10000)
+            return tile.data, (tile[0] != 0) * 255
+        else:
+            return None, None
+
+    def feed(self, tile):
+        """Add data to tile."""
+        self.tile.append(tile)
+
+class MaxIndexMethod(MosaicMethodBase):
```

Comment:
I feel this method is a bit complex. I have used custom methods called `CustomLastBand...` 

The idea is to use `expression` like `B1,B2,B3,{NDVI}` which will create a 4 bands array and then within the mosaic method, we will exclude the last band from the array. 

```python
class CustomLastBandHigh(MosaicMethodBase):
    """Feed the mosaic tile using the last band as decision factor."""

    @property
    def data(self):
        """Return data and mask."""
        if self.tile is not None:
            return self.tile.data[:-1], ~self.tile.mask[0] * 255
        else:
            return None, None

    def feed(self, tile: numpy.ma.array):
        """Add data to tile."""
        if self.tile is None:
            self.tile = tile
            return

        pidex = (
            numpy.bitwise_and(tile.data[-1] > self.tile.data[-1], ~tile.mask)
            | self.tile.mask
        )

        mask = numpy.where(pidex, tile.mask, self.tile.mask)
        self.tile = numpy.ma.where(pidex, tile, self.tile)
        self.tile.mask = mask


class CustomLastBandLow(MosaicMethodBase):
    """Feed the mosaic tile using the last band as decision factor."""

    @property
    def data(self):
        """Return data and mask."""
        if self.tile is not None:
            return self.tile.data[:-1], ~self.tile.mask[0] * 255
        else:
            return None, None

    def feed(self, tile: numpy.ma.array):
        """Add data to tile."""
        if self.tile is None:
            self.tile = tile
            return

        pidex = (
            numpy.bitwise_and(tile.data[-1] < self.tile.data[-1], ~tile.mask)
            | self.tile.mask
        )

        mask = numpy.where(pidex, tile.mask, self.tile.mask)
        self.tile = numpy.ma.where(pidex, tile, self.tile)
        self.tile.mask = mask
```

Reply:
It makes a lot of sense.

The reason I didn't go for a shipped selection band as part of the asset is in the case of feeding STAC assets directly to `mosaic_reader` from an S3 bucket, this selection criterium band is not available, and that would require computing them upfront. 

I guess this could be done in the `tiler` method passed to `mosaic_reader`, as in example below (pseudo-code to illustrate the idea, I haven't tested it):

```
from satsearch import Search
from rio_tiler.mosaic import mosaic_reader
from rio_tiler.mosaic.methods import defaults
from rasterio.crs import CRS

def tiler(src_path: str, *args, **kwargs) -> Tuple[np.ndarray, np.ndarray]:
    with S2COGReader(src_path) as cog:
        cog_src =  cog.part(*args, **kwargs)
        ndvi = (cog_src[-1] - cog_src[-2]) / (cog_src[-1] - cog_src[-2])
    return np.concatenate([cog_src, ndvi], axis=0)

search = Search(bbox=bbox,
               url="https://earth-search.aws.element84.com/v0",
               datetime=f'{time_interval[0]}/{time_interval[1]}',
               query={'eo:cloud_cover': {'lt': 90}},
               collections=['sentinel-s2-l2a-cogs'])

sentinel, assets_used = mosaic_reader(assets, tiler, bbox,
                                      bounds_crs=CRS.from_epsg(bbox.crs),
                                      pixel_selection=defaults.CustomLastBandHigh(),
                                      bands=('B02', 'B03', 'B04', 'B08'))
```

The problem is that we lose the string reference to the bands in the `tiler` method, which is not really an issue as it is custom set by the user, so it can always be aligned with the bands set in `mosaic_reader`.

Reply:
@wouellette I think you could use `expression` for what you are trying to do.

```python
from satsearch import Search
from rio_tiler_pds.sentinel.aws import S2COGReader
from rio_tiler.mosaic import mosaic_reader
from rio_tiler.mosaic.methods import defaults
from rasterio.crs import CRS

def tiler(src_path: str, *args, **kwargs) -> Tuple[np.ndarray, np.ndarray]:
    with S2COGReader(src_path) as cog:
        return cog.part(*args, **kwargs)

search = Search(
    bbox=bbox,
    url="https://earth-search.aws.element84.com/v0",
    datetime=f'{time_interval[0]}/{time_interval[1]}',
    query={'eo:cloud_cover': {'lt': 90}},
    collections=['sentinel-s2-l2a-cogs']
)

(data, mask), assets_used = mosaic_reader(
    assets, tiler, bbox,
    bounds_crs=CRS.from_epsg(bbox.crs),
    pixel_selection=defaults.CustomLastBandHigh(),
    expression="B02,B03,B04,(B04-B08/B04+B08)"
)

# Note: data would be a (3xnxn) array of type float because of the NDVI band
```

Reply:
I completely overlooked the fact that an expression field was available...

That's sorted then! What's the plan with this? Crystallize these methods in the library? Or was your idea to keep them separate.